File: clients/php/Model/Zone.php
<?php

namespace Model {
    require_once 'Model/Vec2.php';
    require_once 'Stream.php';

    /**
     * Current state of the game zone
     */
    class Zone
    {
        /**
         * Current center
         */
        public \Model\Vec2 $currentCenter;
        /**
         * Current radius
         */
        public float $currentRadius;
        /**
         * Next center
         */
        public \Model\Vec2 $nextCenter;
        /**
         * Next radius
         */
        public float $nextRadius;
    
        function __construct(\Model\Vec2 $currentCenter, float $currentRadius, \Model\Vec2 $nextCenter, float $nextRadius)
        {
            $this->currentCenter = $currentCenter;
            $this->currentRadius = $currentRadius;
            $this->nextCenter = $nextCenter;
            $this->nextRadius = $nextRadius;
        }
    
        /**
         * Read Zone from input stream
         */
        public static function readFrom(\InputStream $stream): Zone
        {
            $currentCenter = \Model\Vec2::readFrom($stream);
            $currentRadius = $stream->readDouble();
            $nextCenter = \Model\Vec2::readFrom($stream);
            $nextRadius = $stream->readDouble();
            return new Zone($currentCenter, $currentRadius, $nextCenter, $nextRadius);
        }
        
        /**
         * Write Zone to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $this->currentCenter->writeTo($stream);
            $stream->writeDouble($this->currentRadius);
            $this->nextCenter->writeTo($stream);
            $stream->writeDouble($this->nextRadius);
        }
    }
}

File: clients/go/pkg/model/zone.go
package model

import (
	"fmt"
	"io"

	"aicup22/pkg/stream"
)

// Current state of the game zone
type Zone struct {
	// Current center
	CurrentCenter Vec2
	// Current radius
	CurrentRadius float64
	// Next center
	NextCenter Vec2
	// Next radius
	NextRadius float64
}

func NewZone(currentCenter Vec2, currentRadius float64, nextCenter Vec2, nextRadius float64) Zone {
	return Zone{
		CurrentCenter: currentCenter,
		CurrentRadius: currentRadius,
		NextCenter:    nextCenter,
		NextRadius:    nextRadius,
	}
}

// Read Zone from reader
func ReadZone(reader io.Reader) Zone {
	currentCenter := ReadVec2(reader)
	currentRadius := stream.ReadFloat64(reader)
	nextCenter := ReadVec2(reader)
	nextRadius := stream.ReadFloat64(reader)
	return Zone{
		CurrentCenter: currentCenter,
		CurrentRadius: currentRadius,
		NextCenter:    nextCenter,
		NextRadius:    nextRadius,
	}
}

// Write Zone to writer
func (zone Zone) Write(writer io.Writer) {
	currentCenter := zone.CurrentCenter
	currentCenter.Write(writer)
	currentRadius := zone.CurrentRadius
	stream.WriteFloat64(writer, currentRadius)
	nextCenter := zone.NextCenter
	nextCenter.Write(writer)
	nextRadius := zone.NextRadius
	stream.WriteFloat64(writer, nextRadius)
}

// Get string representation of Zone
func (zone Zone) String() string {
	stringResult := "{ "
	stringResult += "CurrentCenter: "
	currentCenter := zone.CurrentCenter
	stringResult += currentCenter.String()
	stringResult += ", "
	stringResult += "CurrentRadius: "
	currentRadius := zone.CurrentRadius
	stringResult += fmt.Sprint(currentRadius)
	stringResult += ", "
	stringResult += "NextCenter: "
	nextCenter := zone.NextCenter
	stringResult += nextCenter.String()
	stringResult += ", "
	stringResult += "NextRadius: "
	nextRadius := zone.NextRadius
	stringResult += fmt.Sprint(nextRadius)
	stringResult += " }"
	return stringResult
}
